lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+            $element = $element->parentNode;
+
+            if (! $element instanceof DOMElement) {
+                continue;
+            }
+
+            $width  = $element->getAttribute(Attribute::WIDTH);
+            $height = $element->getAttribute(Attribute::HEIGHT);
+
+            if (! $width && ! $height) {
+                continue;
+            }
+
+            return [$width, $height];
+        }
+
+        return [0, 0];
+    }
+
+    /**
+     * Check whether an element is to be considered tiny and should be ignored.
+     *
+     * A tiny element is any element with width or height less than 150 pixels and a non-responsive layout.
+     *
+     * @param string $layout Layout of the element.
+     * @param int    $width  Width of the element.
+     * @param int    $height Height of the element.
+     * @return bool Whether the element is tiny.
+     */
+    private function isTinyElement($layout, $width, $height)
+    {
+        if ($width <= 0 || $height <= 0) {
+            return true;
+        }
+
+        if ($layout === Layout::INTRINSIC || $layout === Layout::RESPONSIVE) {
+            return false;
+        }
+
+        return $width < self::TINY_IMG_THRESHOLD || $height < self::TINY_IMG_THRESHOLD;
+    }
+
+    /**
+     * Detect a hero image candidate from a video's poster (= placeholder) image.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidateForPosterImage(DOMElement $element)
+    {
+        $poster = $element->getAttribute(Attribute::POSTER);
+
+        if (! $poster) {
+            return null;
+        }
+
+        if (! Url::isValidImageSrc($poster)) {
+            return null;
+        }
+
+        $width  = $element->getAttribute(Attribute::WIDTH);
+        $height = $element->getAttribute(Attribute::HEIGHT);
+        $layout = $element->getAttribute(Attribute::LAYOUT);
+
+        if ($this->isTinyElement($layout, $width, $height)) {
+            return null;
+        }
+
+        $media = $element->getAttribute(Attribute::MEDIA);
+
+        return new HeroImage($poster, $media, '');
+    }
+
+    /**
+     * Detect a hero image candidate from an iframe's placeholder image.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidateForIframePlaceholderImage(DOMElement $element)
+    {
+        // A placeholder <amp-img> is required to preload an image for an iframe.
+        if (! $element->hasChildNodes()) {
+            return null;
+        }
+
+        $width  = $element->getAttribute(Attribute::WIDTH);
+        $height = $element->getAttribute(Attribute::HEIGHT);
+        $layout = $element->getAttribute(Attribute::LAYOUT);
+
+        if ($this->isTinyElement($layout, $width, $height)) {
```

Comment:
Again, `layout` could be `fixed-height`. In such case, the `$width` can be assumed to be large.